feat(toast): add toast package (#760)
twilio-labs/paste: packages/paste-website/src/pages/components/toast/index.mdx

Use a toast for one of these types of responses:

- **Neutral or general response.** Let a user know that an asynchronous process has been triggered based on their action, especially if they have to wait for the process to finish before continuing a workflow. This could include a large download or a request that's been sent to a third party.
- **Success.** Confirm a successful or completed action.
- **Warning.** Help a user avoid a potential future issue that might happen based on their action. If the user can take an immediate action to avoid the issue, provide a way to do that. In most cases, a warning toast lets users know that their current action completed but may have triggered a related issue, or that their action can't be completed because of an issue that they need to fix outside their current flow.
- **Error.** Communicate an issue that's preventing a user from completing their current action and can be fixed within their current flow.

In general, use Toasts for response to user action. To decide whether you should use Toasts or another type of messaging, jump to [Toasts vs. Alerts vs. Inline messaging](#toast-vs-alerts-vs-inline-messaging).

### Accessibility

> For [toasts that automatically time out](#toast-with-automatic-time-out), consider users with different reading speeds, vision levels, and literacy levels. We recommend giving users about 0.5 seconds per word. For short toasts at 5–10 words, that's about 2–4 seconds until you time-out a toast. For toasts at our [max suggested length of 140 characters](#composition-notes), that's about 30 seconds for about 55 English words. In this case, consider not making the toast automatically time out since the longer content length could also suggest that there's information worth persisting for the customer.

Reviewer: ```suggestion
For [toasts that automatically time out](#toast-with-automatic-time-out), consider users with different reading speeds, vision levels, and literacy levels (more details in the research insight below). We recommend giving users about 0.5 seconds per word. For short toasts at 5–10 words, that's about 2–4 seconds until you time-out a toast. For toasts at our [max suggested length of 140 characters](#composition-notes), that's about 30 seconds for about 55 English words. In this case, consider not making the toast automatically time out since the longer content length could also suggest that there's information worth persisting for the customer.
```

Also consider the time users need to make a decision based on the message. If a toast has any interactive element in it, like a link or button, don't add an automatic time-out to the toast.

<Callout>
  <CalloutTitle>Research insight into reading speeds</CalloutTitle>
  <CalloutText>
    From the{' '}
    <Anchor href="https://www.perkinselearning.org/technology/blog/reading-rates">Perkins School for the Blind</Anchor>:
    Reading rates are 1.5–2 times slower for people with low vision compared to those with sighted vision. If we take an
    8th grade reading level at 197 words per minute (wpm), we can suggest a 131 wpm reading rate.
  </CalloutText>
  <CalloutText>
    Additional research by{' '}
    <Anchor href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5557829/">
      researchers at the University of Alabama at Birmingham
    </Anchor>{' '}
    and by{' '}
    <Anchor href="https://www.microsoft.com/en-us/research/uploads/prod/2019/01/readerview_chi2019.pdf">
      Microsoft Research
    </Anchor>{' '}

>     suggest roughly similar or higher reading rates due to increased age and dyslexia in adults.

Reviewer: ```suggestion
    suggest reading rates are similar or slightly faster with increased age or dyslexia in adults.
```

  </Text>
</Toast>`}
</LivePreview>

### Toast with automatic time-out

Add an automatic time-out to a toast if it doesn't communicate any information that would help a user continue a flow, and would be annoying to the user to have to explicitly acknowledge and close.

In most cases, it's better to dismiss the toast when a user reloads or navigates away from a page, rather than timing it out. Add an automatic time-out to a success toasts if the only thing it does it confirm that the action was completed. If the toast has any additional information or interactive content, like a link or button, don't make the toast time out.

Add an automatic time-out if all these conditions are true:

- The user wouldn't need to remember or copy the information in order to continue with their current flow, or any future flow.
- The toast doesn't have any interactive content, such as a link or button.

We recommend giving users about 0.5 seconds per word before timing out a toast.

### Showing and Hiding a Toast

For a basic method of hiding and showing, you can use the `hidden` HTML attribute. Passing a true or false value to that attribute will show or hide the Toast respectively.

> We'll be providing a more comprehensive method for displaying Toasts in the near future.

Reviewer 2: I'd be tempted to make this a callout at the start of this section to make it _really_ clear that we acknowledge this is just an interim solution.

Author: I think thats a good call. Done.

  <Text as="span">
    Your function is currently being deployed.
  </Text>
</Toast>`}
</LivePreview>

### Success Toast

Use a success toast to confirm a successful or completed action. If a creation flow is a 2-step process—a user creates an object, then edits the details of that object—confirm the completion after each step with a toast.

Provide an "undo" action in the toast if it's easily and safely reversible, and you're able to take the user back to their previous state. For example, you can allow the user to undo an action through a toast if:

- They submitted a short form that updates something. Undoing the action can take them back to their filled form with any sensitive fields cleared.

- They deleted non-sensitive information we'd want them to easily retrieve.

<LivePreview scope={{Toast, Text, Anchor}} language="jsx">
  {`<Toast onDismiss={() => alert('dismissed')} variant="success">
  <Text as="span">
    Your email address has been updated. You can view your profile <Anchor href="#">here</Anchor>.
  </Text>
</Toast>`}
</LivePreview>

<Callout>
  <CalloutTitle>When to automatically time-out a success toast</CalloutTitle>
  <CalloutText>

>     Add an automatic time-out to a success toast if the _only_ thing it does is confirm that the action was completed. If

Reviewer: whoops, this is showing markdown syntax. Forgot this was in the callout

<Box marginBottom="space80">
  <Stack orientation="vertical" spacing="space40">
    <Toast onDismiss={() => alert('dismissed')} variant="neutral">
      <Text as="span">
        <strong>Martin Luther King, Jr.:</strong> We may have all come on different ships, but we’re in the same boat
        now.
      </Text>
    </Toast>
    <Toast onDismiss={() => alert('dismissed')} variant="warning">
      <Text as="span">
        <strong>James Baldwin:</strong> Love takes off masks that we fear we cannot live without and know we cannot live
        within.
      </Text>
    </Toast>
    <Toast onDismiss={() => alert('dismissed')} variant="error">
      <Text as="span">
        <strong>Chimamanda Ngozi Adichie:</strong> Racism should never have happened and so you don't get a cookie for
        reducing it.
      </Text>
    </Toast>
    <Toast onDismiss={() => alert('dismissed')} variant="success">
      <Text as="span">
        <strong>Toni Morrison:</strong> You wanna fly, you got to give up the shit that weighs you down.
      </Text>
    </Toast>
  </Stack>
</Box>

### Showing and Hiding a Toast

> <Callout>
>   <CalloutText>
>     We'll be providing a more comprehensive method for displaying Toasts in the near future. Please bare with us.
>   </CalloutText>
> </Callout>

Reviewer: ```suggestion
<Callout variant="warning">
  <CalloutText>
    We'll be providing a more comprehensive method for displaying Toasts in the near future. Please bear with us.
  </CalloutText>
</Callout>
```

> <Callout>
>   <CalloutText>
>     For a basic method of hiding and showing, you can use the <inlineCode>hidden</inlineCode> HTML attribute. Passing a
>     <inlineCode>true</inlineCode> or <inlineCode>false</inlineCode> value to that attribute will show or hide the Toast
>     respectively. We'll be providing a more comprehensive method for displaying Toasts in the near future.
>   </CalloutText>
> </Callout>

Reviewer 2: ```suggestion
<Callout>
  <CalloutText>
    We'll be providing a more comprehensive method for displaying Toasts in the near future. Please bare with us.
  </CalloutText>
</Callout>

For a basic method of hiding and showing, you can use the `hidden` HTML attribute. Passing a `true` or `false` value to that attribute will show or hide the Toast respectively.
```

<LivePreview
  scope={{
    Toast,
    ToastContainer,
    Text,
    Button,
  }}
  noInline
  language="jsx"
>
  {showHideExample}

> </LivePreview>

Author: Let's backlog that, so we can get this package out this sprint.

Be cautious when using a warning toast because they can be stressful for a user to see. If the information you're trying to convey is not very critical, consider using a neutral toast instead.

<LivePreview scope={{Toast, Text, Anchor}} language="jsx">
  {`<Toast onDismiss={() => alert('dismissed')} variant="warning">
  <Text as="span">
    The phone number has been deleted. Please verify there are no functions tied to that number <Anchor href="#">here</Anchor>.
  </Text>
</Toast>`}
</LivePreview>

### Error Toast

Use an error toast to help a user understand an issue that's preventing them from completing their current action. Provide a way to resolve the issue either in the underlying page content or in the toast.

Be cautious when using an error alert because they can be stressful for a user to see.

<LivePreview scope={{Toast, Text}} language="jsx">
  {`<Toast onDismiss={() => alert('dismissed')} variant="error">
  <Text as="span">
    There was an error when deleting the profile. Please try again.
  </Text>
</Toast>`}
</LivePreview>

### Toast with automatic time-out

> Add an automatic time-out to a toast if it doesn't communicate any information that would help a user continue a flow, and would be annoying to the user to have to explicitly acknowledge and close.
> 
> In most cases, it's better to dismiss the toast when a user reloads or navigates away from a page, rather than timing it out. Add an automatic time-out to a success toasts if the only thing it does it confirm that the action was completed. If the toast has any additional information or interactive content, like a link or button, don't make the toast time out.
> 
> Add an automatic time-out if all these conditions are true:
> 
> - The user wouldn't need to remember or copy the information in order to continue with their current flow, or any future flow.
> - The toast doesn't have any interactive content, such as a link or button.
> 
> We recommend giving users about 0.5 seconds per word before timing out a toast.

Reviewer: ```suggestion
Add an automatic time-out if all these conditions are true:

- The user wouldn't need to remember or copy the information in order to continue with their current flow, or any future flow.
- The toast doesn't have any interactive content, such as a link or button.

We recommend giving users about 0.5 seconds per word before timing out a toast.
```

> We recommend giving users about 0.5 seconds per word before timing out a toast.

Author: Yeah. It's a bit of manual process for downstream teams to do that now. They'll have to add some sort of time to the state of the app, and under-render the Toast when the timer runs out. Eventually the Toaster will handle this as well. Let's backlog the example for now so we can get this out the door.

## Composition Notes

Keep toast text brief and scannable by placing only the highest priority information in it. Too much text in a toast can overwhelm a user. Aim for no more than 140 characters in a toast, and keep the text directly related to the user's current action. Success toasts generally run about 50 characters long, while all other alerts might run about 100 characters long. However, don't [sacrifice being direct and straightforward](/content/product-style-guide#be-direct-and-straightforward) for a shorter message.

> Be specific in your response to the user's action and if applicable, provide a resolution by using an [anchor](/components/anchor/). Be cautious about using a button in a toast since it may compete with other buttons on the screen and should be used only to trigger an action. If you must use a button, use a single [small secondary button](/components/button#small-button) in a toast.

Reviewer: ```suggestion
Be specific in your response to the user's action and if applicable, provide a resolution by navigating a user to another page using an [anchor](/components/anchor/). Be cautious about using a button in a toast since it should be used only to trigger an action. If you must use a button, use a [small secondary button](/components/button#small-button) in a toast.
```

> <Callout variant="warning">
>   <CalloutTitle>Don't forget the ToastContainer</CalloutTitle>
>   <CalloutText>
>     In order to control placement for Toasts, Paste provides you with a ToastContainer. The ToastContainer only handles
>     the positioning of Toasts in the top right area of the application. Multiple Toasts can be added to a single
>     ToastContainer. This ensures the correct spacing is added to each Toast.
>   </CalloutText>
> </Callout>

Reviewer: ```suggestion
<Callout>
  <CalloutTitle>Don't forget the ToastContainer!</CalloutTitle>
  <CalloutText>
    In order to control placement for Toasts, Paste provides you with a ToastContainer. The ToastContainer only handles
    the positioning of Toasts in the top right area of the application. Multiple Toasts can be added to a single
    ToastContainer. This ensures the correct spacing is added to each Toast.
  </CalloutText>
</Callout>
```

## When to use a Toast

> Help users avoid scenarios where more than 3 toasts could show at a time. If more than 3 could appear at a time, show only one toast that summarizes the information you were intending to show in multiple toasts.

Reviewer: ```suggestion
Help users avoid scenarios where more than 3 toasts could show at a time. If more than 3 could appear at a time, show only one toast that summarizes the information you were intending to show in all of the toasts.
```

### Toast vs. Alerts vs. Inline messaging

To communicate feedback on a user action, use a toast or in-page validation (_e.g._, full-page confirmation, inline error). You may also want to pair a toast with in-page validation. For example, when a user submits a long form, you can show an error toast that lets users know they need to fix some form fields, while also showing inline errors on each incorrect field.

To communicate information from the system that isn't triggered by a user action, use an [alert](/components/alert) or other in-page messaging.

##### Use Toasts when...

> - You need to communicate a direct response to a user action. **Example:** "asset.jpg" successfully added.

Reviewer: ```suggestion
- You need to communicate a direct response to a user action. Example: _"asset.jpg" successfully added._
```

- You're providing only one link or button to help a user continue a flow.

##### Use Alerts when...

- You need to notify a user about a high-priority or time-sensitive change that isn't triggered by a user action.
- You're providing only one link or button to help a user resolve an issue.

> ##### Use in-page messaging e.g., inline errors, in-page call-outs when...

Reviewer: ```suggestion
##### Use in-page messaging, e.g., inline errors, in-page call-outs when...
```

- You need to communicate a direct and contextual response to a user action.

> - You need to provide links or buttons contextually to an in-page element.

Reviewer: ```suggestion
- You need to provide links or buttons next to an in-page element.
```

> - You need to communicate dynamically changing content inside a single toast, including showing progress bars. These cases should use a non-modal dialog instead.

Reviewer: ```suggestion
- You need to communicate dynamically changing content, like a progress bar. Use a [non-modal dialog](primitives/non-modal-dialog-primitive) instead.
```
